openedx/core/djangoapps/ccxcon/api.py

```diff
@@ -0,0 +1,146 @@
+from instructor.access import list_with_level
+from openedx.core.djangoapps.models.course_details import CourseDetails
+from student.models import anonymous_id_for_user
+from .models import CCXCon
+
+log = logging.getLogger(__name__)
+
+
+CCXCON_COURSEXS_URL = '/api/v1/coursexs/'
+
+
+class CCXConnServerError(Exception):
+    """
+    Custom exception to be raised in case there is any
+    issue with the request to the server
+    """
+
+
+def is_valid_url(url):
+    """
+    Helper function used to check if a string is a valid url.
+
+    Args:
+        url (str): the url string to be validated
+
+    Returns:
+        bool: whether the url is valid or not
+    """
+    validate = URLValidator()
+    try:
+        validate(url)
+        return True
+    except ValidationError:
+        return False
+
+
+def get_oauth_client(server_token_url, client_id, client_secret):
+    """
+    Function that creates an oauth client and fetches a token.
+    It intentionally doesn't handle errors.
+
+    Args:
+        server_token_url (str): server URL where to get an authentication token
+        client_id (str): oauth client ID
+        client_secret (str): oauth client secret
+
+    Returns:
+        an instance of OAuth2Session with a token
+    """
+    if not is_valid_url(server_token_url):
+        return
+    client = BackendApplicationClient(client_id=client_id)
+    oauth_ccxcon = OAuth2Session(client=client)
+    oauth_ccxcon.fetch_token(
+        token_url=server_token_url,
+        client_id=client_id,
+        client_secret=client_secret,
+        verify=False
+    )
+    return oauth_ccxcon
+
+
+def course_info_to_ccxcon(course_key):
+    """
+    Function that gathers informations about the course and
+    makes a post request to a CCXCon with the data.
+
+    Args:
+        course_key (CourseLocator): the master course key
+    """
+
+    try:
+        course = get_course_by_id(course_key)
+    except Http404:
+        log.error('Master Course with key "%s" not found', unicode(course_key))
+        return
+    if not course.enable_ccx:
+        log.debug('ccx not enabled for course key "%s"', unicode(course_key))
+        return
+    if not course.ccx_connector:
+        log.debug('ccx connector not defined for course key "%s"', unicode(course_key))
+        return
+    if not is_valid_url(course.ccx_connector):
+        log.error(
+            'ccx connector URL "%s" for course key "%s" is not a valid URL.',
+            course.ccx_connector, unicode(course_key)
+        )
+        return
+    # get the oauth credential for this URL
+    try:
+        ccxcon = CCXCon.objects.get(url=course.ccx_connector)
+    except CCXCon.DoesNotExist:
+        log.error('ccx connector Oauth credentials not configured for URL "%s".', course.ccx_connector)
+        return
+
+    # get an oauth client with a valid token
+    oauth_ccxcon = get_oauth_client(
+        server_token_url=urlparse.urljoin(course.ccx_connector, '/o/token/'),
```

Comment:
that became a global variable because I needed to import it in the tests, but I'll make also this one global